sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -491,6 +493,9 @@ private[sql] object DataSourceStrategy extends Strategy with Logging {
     // Translated data source filters that can be handled by `relation`
     val (_, handledFilters) = handled.unzip
 
-    (unrecognizedPredicates ++ unhandledPredicates, handledFilters)
+    // translated contains all filters that have been converted to the public Filter interface.
+    // We should always push them to the data source no matter whether the data source can apply
+    // a filter to every row or not.
+    (unrecognizedPredicates ++ unhandledPredicates, translated.map(_._2))
```

Comment:
Would be nice to make `translated.map(_._2)` more explicit and readable as follows, since this part is particularly tricky to understand:

```scala
// Translated data source filters, no matter `relation` can handle them or not
val (_, translatedFilters) = translated.unzip

// ...
(unrecognizedPredicates ++ unhandledPredicates, translatedFilters)
```